configure: use variables rather than arguments for a few options

shindere opened the pull request:
This PR replaces a few configure command-line options by configuration
variables:

- "--with-dllibs" is replaced by DLLIBS
- "--with-reserved-header-bits" is replaced by RESERVED_HEADER_BITS
- "--with-default-string" is replaced by DEFAULT_STRING

dra27 commented:
Thoughts and cross-references to the original PR!

cf. https://github.com/ocaml/ocaml/pull/2139#discussion_r240035814 which suggested using `AC_ARG_ENABLE` for the string mode. I agree(d) with the `DLLIBS` change (cf. https://github.com/ocaml/ocaml/pull/2139#discussion_r240035855). For `--with-reserved-header-bits` was noted in https://github.com/ocaml/ocaml/pull/2139#discussion_r240035661, I suggested switching to `AC_ARG_ENABLE` - I don't have as strong an opinion, but it feels as though `AC_ARG_VAR` should be **always** responsible for changing something (like flags, etc.) and not for enabling/disabling something (i.e. there should not be a special value for an environment which is responsible for turning things on and off).

gasche commented:
Is the idea that environment variables are a better interface than command-line arguments? (I generally tend to assume the converse, that explicit arguments are a clearer interface than passing configuration options on-the-side through the environment; but then I don't have much experience with autoconf conventions etc.)

For `DEFAULT_STRING` I'm not convinced by the idea that force/don't-force is a command-line parameter but safe/unsafe default is in the environment; it makes sense to have all the options at the same place -- possibly as @dra27 suggests as a single setting, `force / default-safe / default-unsafe`, although personally I'm fine with having two different knobs (will comment in the relevant discussion).

mshinwell commented:
I'm not sure what the rationale for this patch is, but I'm against using environment variables for configuration in general.  The problem with environment variables is that their state is easy to forget, for a human user; and for a program calling a script that depends on such variables, you need to take care that they are in the correct state.  That basically means clearing the environment and starting from scratch, which isn't great.

shindere commented:
Mark Shinwell (2019/04/15 04:52 -0700):
> I'm not sure what the rationale for this patch is, but I'm against
> using environment variables for configuration in general.

Beware, you are mixing up two notions here: environemnt varialbe and
configure variables.

The intended use of the configure variables is:

```bash
./configure VARIABLE1=value1 VARIABLE2=value2
```

rather than

```bash
VARIABLE1=value1 VARIABLE2=value2 ./configure
```

whose use is discouraged.

> The problem with environment variables is that their state is easy to
> forget, for a human user; and for a program calling a script that
> depends on such variables, you need to take care that they are in the
> correct state.  That basically means clearing the environment and
> starting from scratch, which isn't great.

Well, beyond likes and dislikes, this use of configuration variables is
the one recommended by the autoconf documentation for all the settings
that have not to do with either enabling an external package
(--with|without-package) or enabling an (internal) feature
(--enable|disable-feature).

shindere commented:
Gabriel Scherer (2019/04/15 04:43 -0700):
> Is the idea that environment variables are a better interface than
> command-line arguments?

No, they are a different one. I think I already explained this a bit in
my response to Mark's comment, but if you need me to further clarify,
please let me know.

> (I generally tend to assume the converse, that explicit arguments are
> a clearer interface than passing configuration options on-the-side
> through the environment; but then I don't have much experience with
> autoconf conventions etc.)

It seems to me that, like Mark, you are mixing environment and
configuration varialbes.

> For `DEFAULT_STRING` I'm not convinced by the idea that force/don't-force is a command-line parameter but safe/unsafe default is in the environment; it makes sense to have all the options at the same place -- possibly as @dra27 suggests as a single setting, `force / default-safe / default-unsafe`, although personally I'm fine with having two different knobs (will comment in the relevant discussion).

Sorry -- would it be possible to discuss here rather than on the
original configure PR? A bit easier to follow for me.

gasche commented:
I had the same confusion as Mark (I wasn't actually aware of "configuration variables" as a distinct thing from "environment variables"). Your explanation justifies having `--enable-force-safe-string` but `DEFAULT_STRING=...`.

Some further questions come from a closer look:
- when is it appropriate to use the `--enable-feature[=ARG]` form that takes an argument?
- what do the `[=PKGS]` parameters mean for `--enable-{shared,static,fast-install}` and `--with-pic`?
- wouldn't `--enable-reserved-header-bits=NUM` make sense? (This is a feature that is enabled?)

I suppose that you are also planning to convert `--with-pic` and `--with-sysroot` to this style later?

gasche commented:
Re force-safe-string vs. default-safe-string, I wrote on the thread in the other PR:

> There is a conceptual difference between:
> 
>  * `--force-safe-string`, which must be a configure-time option that changes the runtime definitions and has a global impact on all modules compiled and linked by the resulting compiler
> 
> * `--default-string={safe,unsafe}` which is "merely" a convenience feature to set a default for compile-time options, rather than a configure-time option by nature (`afl` is the same, etc.). It does nothing more than change the default for a choice that the user makes on a per-invocation basis.

> In my opinion this conceptual difference justifies _not_ showing the three options as a single knob to fix at configure-time, but two separate kind of options. There is a dependency between them (`--force-safe-string` makes it incorrect to change the per-invocation value, so it makes `--default-string` useless and meaningless), but that is common across the configuration space, for example `--disable-native-compiler` will make many other settings meaningless.

shindere commented:
Gabriel Scherer (2019/04/15 14:49 +0000):
> I had the same confusion as Mark (I wasn't actually aware of
> "configuration variables" as a distinct thing from "environment
> variables").

I was not really aware of the distinction either before I had to tinker
with autoconf business.

> Your explanation justifies having `--enable-force-safe-string` but
> `DEFAULT_STRING=...`.

Good, thanks.

> Some further questions come from a closer look:
> - when is it appropriate to use the `--enable-feature[=ARG]` form that
> takes an argument?

Well to give you an example, in brltty there is
``--enable-braille-drivers=DRIVER_CODES`` which takes as argument a list
of driver names to build.

But I couldn't find in autoconf's documentation a more general
explanation, after having looked quickly, sorry.

> - what do the `[=PKGS]` parameters mean for
> - `--enable-{shared,static,fast-install}` and `--with-pic`?

I'm not sure. These options have been provided by libtool so its
documentation may tell you.

> - wouldn't `--enable-reserved-header-bits=NUM` make sense? (This is a
> - feature that is enabled?)

Yes, I think that would make sense. I can implement this if this is fine
with @dra27?

> I suppose that you are also planning to convert `--with-pic` and
> `--with-sysroot` to this style later?

No, that wouldn't actually not be possible, sorry, because the first is
provided by libtool and the second by autoconf itself.

shindere commented:
David Allsopp (2019/04/15 01:50 -0700):
> Thoughts and cross-references to the original PR!
> 
> cf. https://github.com/ocaml/ocaml/pull/2139#discussion_r240035814
> which suggested using `AC_ARG_ENABLE` for the string mode.

So I think I should let this as it is given @gasche's response?

> I agree(d) with the `DLLIBS` change (cf.
> https://github.com/ocaml/ocaml/pull/2139#discussion_r240035855).

Good, so I am not touching this.

> For `--with-reserved-header-bits` was noted in
> https://github.com/ocaml/ocaml/pull/2139#discussion_r240035661, I
> suggested switching to `AC_ARG_ENABLE` - I don't have as strong an
> opinion, but it feels as though `AC_ARG_VAR` should be **always**
> responsible for changing something (like flags, etc.) and not for
> enabling/disabling something (i.e. there should not be a special value
> for an environment which is responsible for turning things on and
> off).

I am fine with that and actually doing it right now.
What should configure do if called with `--enable-reserved-header-bits`,
i.e. the argument is given but with no value?

shindere commented:
Code updated to use `AC_ARG_ENABLE` to define the number of bits
reserved for profiling info in block headers.

Only numbers between 0 and 31 are accepted as arguemnts.

Giving no argument to this option is considered an error.

gasche commented:
> Only numbers between 0 and 31 are accepted as arguemnts.

I'm not sure why you chose 31 as an upper bound? I haven't seen such a bound in the implementation and it seems to me that some choices between 32 and 54 would still make sense.

shindere commented:
Gabriel Scherer (2019/04/16 05:18 -0700):
> > Only numbers between 0 and 31 are accepted as arguemnts.
> 
> I'm not sure why you chose 31 as an upper bound? I haven't seen such a
> bound in the implementation and it seems to me that some choices
> between 32 and 54 would still make sense.

It was 32 in the previous implementation (see e.g. the configure script
for 4.07) and the help said
the number should be less than 32 (and not less or equal) so I used
31 but I can change this if necessary.

gasche commented:
I guess 31 or 32 is fine then. (I would find it strange if there is nothing closer to the actual logic of the code that checks that this parameter is sensible, but you're not responsible for checking or changing this.)

shindere commented:
Gabriel Scherer (2019/04/16 05:32 -0700):
> I guess 31 or 32 is fine then. (I would find it strange if there is
> nothing closer to the actual logic of the code that checks that this
> parameter is sensible, but you're not responsible for checking or
> changing this.)

Okay great.

Is there something missing from the PR then? @dra27? @gasche? is the
current way of dealing with string-related options okay for both of you,
ifnally?

damiendoligez reviewed (approved):
The code looks good and it seems everyone agrees on the interface.

shindere commented:
OK thanks @damiendoligez.
Rebasing and merging once CI is happy.

damiendoligez commented on configure.ac:
```diff
@@ -102,7 +102,6 @@ AC_SUBST([oc_cflags])
 AC_SUBST([oc_cppflags])
 AC_SUBST([oc_ldflags])
 AC_SUBST([bytecclibs])
-AC_SUBST([dllibs])
```
I would naively expect this to be replaced by `AC_SUBST([DLLIBS])` rather than removed altogether.

damiendoligez commented on configure.ac:
```diff
@@ -1265,7 +1267,7 @@ AS_IF([$shared_libraries_supported],
     [AC_CHECK_FUNC([dlopen],
       [supports_shared_libraries=true dllibs=""],
```
Here `dllibs` should be replaced with `DLLIBS`, right?

damiendoligez reviewed (commented):
CI is unhappy because of something to do with Dynlink.

shindere committed:
configure: use variables rather than arguments for a few options

This commit replaces a few configure command-line options by configuration
variables:

- "--with-dllibs" is replaced by DLLIBS
- "--with-reserved-header-bits" is replaced by RESERVED_HEADER_BITS
- "--with-default-string" is replaced by DEFAULT_STRING

shindere commented:
Damien Doligez (2019/05/07 14:40 +0000):
> damiendoligez commented on this pull request.
> 
> CI is unhappy

Yep, seen, thanks

> because of something to do with Dynlink.

Oh thanks for having investigated!

> > @@ -102,7 +102,6 @@ AC_SUBST([oc_cflags])
>  AC_SUBST([oc_cppflags])
>  AC_SUBST([oc_ldflags])
>  AC_SUBST([bytecclibs])
> -AC_SUBST([dllibs])
> 
> I would naively expect this to be replaced by `AC_SUBST([DLLIBS])`
> rather than removed altogether.

No in fact when you use something like

```
AC_ARG_VAR([DLLIBS],...)
```

then `DLLIBS` gets substituded automatically so you odn't need to be
explicit about it.

> > @@ -1265,7 +1267,7 @@ AS_IF([$shared_libraries_supported],
>      [AC_CHECK_FUNC([dlopen],
>        [supports_shared_libraries=true dllibs=""],
> 
> Here `dllibs` should be replaced with `DLLIBS`, right?

Yes, absolutely.

And in a few other places also, actually.

Updating the branch and let's see whether this restores CI's happiness.

shindere commented:
Merged into trunk and cherry-picked to 4.08 as commit
6da973d14f3ecbe6d7c0fb17f6ec8acc1786c9b1

trefis commented:
It would be a shame if you forgot about 4.09.

shindere commented:
Thomas Refis (2019/05/09 08:37 -0700):
> It would be a shame if you forgot about 4.09.

Oh damn, it's true that we now have two release branches, many thanks
for the reminder @trefis!

So this has been cherry-picked to 4.09 as commit
84ee761ec4a3fb248bb1493293d7b5781f90f079
